### PR TITLE
traffic schema

### PR DESCRIPTION
**Pull request by Tan4ek**

### CODE REVIEW

File: migrations/versions/2022-04-16_56001dbb442e_traffic_model.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+# revision identifiers, used by Alembic.
+revision = '56001dbb442e'
+down_revision = 'cd4c37ebabe2'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_table('traffic_record',
```
Comment (TimAle) [inline migrations/versions/2022-04-16_56001dbb442e_traffic_model.py]: make all the table names plural if these names relate to a single row content

File: migrations/versions/2022-04-16_56001dbb442e_traffic_model.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+                    sa.Column('date_to', sa.Integer(), nullable=False),
+                    sa.Column('provider_id', sa.String(), nullable=False),
+                    sa.Column('quantity_bytes', sa.Integer(), nullable=False),
+                    sa.ForeignKeyConstraint(['provider_id'], ['provider.id'], ),
+                    sa.PrimaryKeyConstraint('id')
+                    )
+    op.create_index('idx_from_to', 'traffic_record', ['date_from', 'date_to'], unique=False)
+    op.execute("ALTER TABLE provider ADD COLUMN external_id VARCHAR NOT NULL DEFAULT ''")
+    op.execute("UPDATE provider SET external_id = substr(payload,20,36) WHERE type='openvpn'")
+    op.create_index(op.f('ix_provider_external_id'), 'provider', ['external_id'], unique=True)
```
Comment (TimAle) [inline migrations/versions/2022-04-16_56001dbb442e_traffic_model.py]: probably `idx_provider_external_id`

File: migrations/versions/2022-04-16_56001dbb442e_traffic_model.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+                    sa.Column('date_from', sa.Integer(), nullable=False),
+                    sa.Column('date_to', sa.Integer(), nullable=False),
+                    sa.Column('provider_id', sa.String(), nullable=False),
+                    sa.Column('quantity_bytes', sa.Integer(), nullable=False),
+                    sa.ForeignKeyConstraint(['provider_id'], ['provider.id'], ),
+                    sa.PrimaryKeyConstraint('id')
+                    )
+    op.create_index('idx_from_to', 'traffic_record', ['date_from', 'date_to'], unique=False)
+    op.execute("ALTER TABLE provider ADD COLUMN external_id VARCHAR NOT NULL DEFAULT ''")
+    op.execute("UPDATE provider SET external_id = substr(payload,20,36) WHERE type='openvpn'")
```
Comment (TimAle) [inline migrations/versions/2022-04-16_56001dbb442e_traffic_model.py]: nasty boy!

File: repository/persistant.py [hunk 4]
```diff
@@ -26,9 +26,22 @@ class VpnProviderModel(Base):
...
     payload = Column("payload", JSON)
+    external_id = Column("external_id", String, unique=True, index=True, nullable=False, default=str(uuid.uuid4()))
     created_at = Column("created_at", DateTime, nullable=False)
 
 
+class TrafficRecordModel(Base):
+    __tablename__ = "traffic_record"
+    id = Column("id", Integer, primary_key=True, autoincrement=True)
+    date_from = Column("date_from", Integer, nullable=False)
+    date_to = Column("date_to", Integer, nullable=False)
```
Comment (TimAle) [inline repository/persistant.py:37]: any chances we can have it as a date?

File: repository/persistant.py [hunk 5]
```diff
@@ -102,12 +115,14 @@ def get_codes(self) -> List[InviteCodeEntity]:
         return [InviteCodeEntity(model.code, model.description) for model in
                 (self._session.query(InviteCodeModel).order_by(InviteCodeModel.created_at).all())]
 
-    def create_provider(self, invite_code: str, provider_type: str, payload: str) -> VpnProviderEntity:
+    def create_provider(self, invite_code: str, external_id: str, provider_type: str,
+                        payload: str) -> VpnProviderEntity:
```
Comment (TimAle) [inline repository/persistant.py:119]: looks like this method should create not a VPN provider only, but a provider in general

**Review by TimAle — CHANGES_REQUESTED** [review]
`persistant.py` -> `persistent.py`

File: migrations/versions/2022-04-16_56001dbb442e_traffic_model.py (hunk 1 above)
Comment (Tan4ek) [reply migrations/versions/2022-04-16_56001dbb442e_traffic_model.py]: Thanks, renamed ✅

File: migrations/versions/2022-04-16_56001dbb442e_traffic_model.py (hunk 3 above)
Comment (Tan4ek) [reply migrations/versions/2022-04-16_56001dbb442e_traffic_model.py]: 🤷‍♂️

File: migrations/versions/2022-04-16_56001dbb442e_traffic_model.py (hunk 2 above)
Comment (Tan4ek) [reply migrations/versions/2022-04-16_56001dbb442e_traffic_model.py]: Thanks ✅

File: repository/persistant.py (hunk 4 above)
Comment (Tan4ek) [reply repository/persistant.py:37]: No, sqlite doesn't have native date data type https://www.sqlite.org/datatype3.html#date_and_time_datatype . Here I use integer approach

File: repository/persistant.py (hunk 5 above)
Comment (Tan4ek) [reply repository/persistant.py:119]: Project name is 'get-your-vpn', so I've decided to use vpn in naming to )

**Review by TimAle — APPROVED** [review]

**Pull request merged**